### ISSUE TITLE
Feature request: allow bitwise on enums

### ISSUE THREAD

**Issue by levibotelho** [palantir/tslint#1266]
#### Actual behavior

`no-bitwise` doesn't allow bitwise operations on enums.
#### Expected behavior

I would like to disallow bitwise operations when dealing with variables, but would like to allow them for enum comparisons (flags), much as they are already allowed for union types.

Is this reasonable?

**Comment by adidahiya** [comment]
sure, this would make sense as a new rule option called `allow-enums`

**Comment by IllusionMH** [comment]
Looks like implementation of this flag might require change base class of this rule from `Lint.Rules.AbstractRule` to `Lint.Rules.TypedRule` since we have to know if we a dealing with enums and not other types. This will be breaking change - this rule is in list of recommended and will require `--type-check` flag after update.

I have an idea how to implement this, but not sure if it is desired change in rule. Probably rule flags might be a better solution.

@adidahiya, @JKillian do we need this changes?

**Comment by JKillian** [comment]
I'm personally against it. For now, my feelings are that we should keep the majority of rules non-type-checked, and especially shouldn't change existing rules to require type-checking.

**Comment by adidahiya** [comment]
+1 to @JKillian's thoughts

**Comment by IllusionMH** [comment]
If you are aware of any other ways to implement this without such breaking change, I'd try to implement it.

As partial solution I can propose two steps:
1. Add option `allow-in-enum-declaration` which will allows next operators **only** inside enum declarations:
- Allow `<<` operator if both of operands are number literals (if I'm right it should be available without type checking)
- Allow `|` operator if both of operands are identifiers of this enum members.

**Example:**
```ts
enum Directions {
    Up = 1 << 0,
    Down = 1 << 1,
    Left = 1 << 2,
    Right = 1 << 3,
    Vertical = Up | Down,
    Horizontal = Left | Right
}
```
**Reasoning:** I think left shift operator might be more expressive and easy to scan than remembering/typing powers of 2. Bitwise or is usable for combinations.
Also I think that it is possible to implement this without converting to `TypedRule`.
Personally I like this idea.

2. Add `skip-bitwise-and` option to rule.
This will allow to delegate typechecks to new rule (something like `enum-bitwice-and`) which will extend `TypedRule` and check if at least one of operands is member of enum.
Also it might check/enforce position (with options like `enum-right` or `enum-left`).
This might be not optimal solutions in general. And I don't know if it is possible(reasonable) to make rules combintations.

(Sidenote: it is ok to mention both of you in questions regarding Issues/PRs or better to just leave a comment without mentions if it isn't direct conversation?)

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is being deprecated** and no longer accepting pull requests for major new changes or features. See #4534. 😱

If you'd like to see this change implemented, you have two choices:
* **Recommended:** Check if this is available in [ESLint](https://eslint.org) + [typescript-eslint](https://typescript-eslint.io) ✅
* _Not Recommended: Fork TSLint locally_ 🤷‍♂️

👋 It was a pleasure open sourcing with you!

_If you believe this message was posted here in error, please comment so we can re-open the issue!_

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
